File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> count = new HashMap();
        //min heap to store (num, freq) pairs ordered by frequency
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap = new PriorityQueue<>(Map.Entry.comparingByValue());
        int[] output = new int[k];
        for (int i: nums) {
            if (count.containsKey(i)) {
                count.put(i, count.get(i) + 1);
            } else {
                count.put(i, 1);
            }
        }
        //iterates through every num, freq pair in the map
        for (Map.Entry<Integer, Integer>entry : count.entrySet()) {
            //adds current num, freq pair to heap
            minHeap.offer(entry);

            if (minHeap.size() > k) {
                //if more than k elements, then smallest freq removed and only k most freq numbers remain
                minHeap.poll();
            }
        }
        
        //removes the k entries from the heap and stores their associated integer keys in the output array
        for (int i = 0; i < k; i++) {
            output[i] = minHeap.poll().getKey();
        }

        return output;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r = 0, len(heights) - 1
        output  = 0
        while l < r:
            total = min(heights[l], heights[r]) * (r-l)
            output = max(output, total)
            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1
        return output